test/unit/mongo_logger.test.ts

```diff
@@ -0,0 +1,204 @@
+      it('only uses the default severity for those components', function () {
+        const loggerOptions = MongoLogger.resolveOptions(
+          {
+            MONGODB_LOG_COMMAND: '',
+            MONGODB_LOG_TOPOLOGY: undefined,
+            MONGODB_LOG_SERVER_SELECTION: 'invalid',
+            MONGODB_LOG_CONNECTION: SeverityLevel.EMERGENCY,
+            MONGODB_LOG_ALL: SeverityLevel.CRITICAL
+          },
+          {}
+        );
+        expect(loggerOptions.command).to.equal(loggerOptions.defaultSeverity);
+        expect(loggerOptions.topology).to.equal(loggerOptions.defaultSeverity);
+        expect(loggerOptions.serverSelection).to.equal(loggerOptions.defaultSeverity);
+        expect(loggerOptions.connection).to.equal(SeverityLevel.EMERGENCY);
+      });
+    });
+
+    context('when environment variables and client options are both set', function () {
+      it('gives precedence to environment variables', function () {
+        const loggerOptions = MongoLogger.resolveOptions(
+          { MONGODB_LOG_PATH: 'env' },
+          { mongodbLogPath: 'client' }
+        );
+        expect(loggerOptions.logDestination).to.equal('env');
+      });
+    });
+
+    context('maxDocumentLength', function () {
+      context('when MONGODB_LOG_MAX_DOCUMENT_LENGTH is undefined', function () {
+        it('defaults to 1000', function () {
+          const loggerOptions = MongoLogger.resolveOptions(
+            {
+              MONGODB_LOG_MAX_DOCUMENT_LENGTH: undefined
+            },
+            {}
+          );
+          expect(loggerOptions.maxDocumentLength).to.equal(1000);
+        });
+      });
+
+      context('when MONGODB_LOG_MAX_DOCUMENT_LENGTH is an empty string', function () {
+        it('defaults to 1000', function () {
+          const loggerOptions = MongoLogger.resolveOptions(
+            {
+              MONGODB_LOG_MAX_DOCUMENT_LENGTH: ''
+            },
+            {}
+          );
+          expect(loggerOptions.maxDocumentLength).to.equal(1000);
+        });
+      });
+
+      context('when MONGODB_LOG_MAX_DOCUMENT_LENGTH cannot be parsed as a uint', function () {
+        it('defaults to 1000', function () {
+          const loggerOptions = MongoLogger.resolveOptions(
+            {
+              MONGODB_LOG_MAX_DOCUMENT_LENGTH: 'invalid'
+            },
+            {}
+          );
+          expect(loggerOptions.maxDocumentLength).to.equal(1000);
+        });
+      });
+
+      context('when MONGODB_LOG_MAX_DOCUMENT_LENGTH can be parsed a uint', function () {
+        it('uses the passed value', function () {
+          const loggerOptions = MongoLogger.resolveOptions(
+            {
+              MONGODB_LOG_MAX_DOCUMENT_LENGTH: '500'
+            },
+            {}
+          );
+          expect(loggerOptions.maxDocumentLength).to.equal(500);
+        });
+      });
+    });
+
+    context('logDestination', function () {
+      context('when mongodbLogPath is undefined', function () {
```

Comment:
I think the contexts should either be split by the value of `mongodbLogPath` or by the value of `MONGODB_LOG_PATH` - mix and matching them in the top level describe is confusing; the tests here would be best served by a table, so that we can tell that we are testing all combinations of interest:
```js
const logDestinationTestCases = [
  {
     description: 'should default to stderr when mongoLogPath and MONGODB_LOG_PATH are undefined',
     MONGODB_LOG_PATH: undefined,
     mongodbLogPath: undefined,
     expectedDestination: 'stderr'
  },
  ...
];

for ... of
   it(testCase.description, () => {
     const loggerOptions = MongoLogger.resolveOptions({ MONGODB_LOG_PATH }, { mongodbLogPath });
     expect(loggerOptions).to.have.property('logDestination', expectedDestination);
   })
```

Reply:
I agree about mixing and matching the top level contexts - but I disagree about the table being easier to read.  I suggested the nested context approach for readability both in tests and test output.  The long, oneline descriptions make the test output harder to grok because the relevant conditions are at the end of each line instead of forming a new section in the output (like they do with nested contexts).  I also don't think the table approach makes it any more clear that we're testing all relevant cases?

Reply:
it's easier to check that we got all the combinations covered by having
```
MONGODB_LOG_PATH: <value>,
mongodbLogPath: <value>,
```
next to each other than by having them buried inside `it` blocks; it's also easier to make sure that the expectation is correct by just checking the `expectedDestination` variable instead of reading the whole expect assertion sentence;
and since the setup is identical for all of these, it's much easier to make sure it stays consistent if we don't replicate it across 9+ blocks;

if the concern is just the shorter lines, those aren't mutually exclusive with this setup, just remove the description variable and generate it from the 3 variables in the test case, then you can do a nested loop and add the `mongodbLogPath` or the env var as a context block and filter the cases on that value

Reply:
if you really wanted to, you could even just have an array like this:
```js
// Format: [ MONGODB_LOG_PATH, [mongodbLogPath], expectedDestination ]
[
  [ undefined, undefined, 'stderr' ],
  [ undefined, 'invalid', 'stderr'],
  ...
]
```

Reply:
we actually are interested in testing a lot of permutations, and that makes the maintainability burden of copy-pasted case instances non-trivial in multiple ways:
1) Updating the code implementation (for example, nesting the logDestination in another object, or changing the format of the input to `resolveOptions`) will require changes in a dozen different places instead of just one
2) Updating the test implementation will require changes in a dozen different places instead of just one
3) Copy pasting is the root of all evil: and it's super easy to miss an incorrectly named test or an incorrect assertion when scanning over a dozen test cases all of which look approximately the same
4) Noticing inconsistencies in implementation is virtually impossible when test cases are broken up across hundreds of lines (e.g., if you decide that the way we treat `undefined` in the logger implementation should be updated, you might not think about the same case for `""` because it's a hundred lines down)

Point 2 above is particularly relevant, because we should use `expect(loggerOptions).to.have.property('logDestination', <value>)` style of assertions throughout this file because we want to have helpful errors. This would be a one line change in a table test because there is a clear separation between the expected input-output expectations of the test and the algorithm that we want to execute to test the input-output combinations.